src/webapi/mutation_observer.rs

```diff
@@ -0,0 +1,259 @@
+use webcore::value::{Reference, Value, ConversionError};
+use webapi::node_list::NodeList;
+use webcore::try_from::{TryFrom, TryInto};
+use webapi::node::{INode, Node};
+
+
+/// `MutationObserver` provides a way to receive notifications about changes in the DOM.
+///
+/// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver)
+pub struct MutationObserver( Reference );
+
+reference_boilerplate! {
+    MutationObserver,
+    instanceof MutationObserver
+}
+
+
+/// Specifies which changes should be observed for the target.
+///
+/// Only used with the [`MutationObserver::observe`](struct.MutationObserver.html#method.observe) method.
+///
+/// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver#MutationObserverInit)
+#[ derive( Debug, Clone ) ]
+pub struct MutationObserverInit< 'a > {
+    /// If `true` it will observe all inserts and removals of the target's children (including text nodes).
+    ///
+    /// This is **not** recursive, it will only observe immediate children
+    /// (unless [`subtree`](struct.MutationObserverInit.html#structfield.subtree) is `true` in which case it will
+    /// observe **all** children and sub-children recursively).
+    pub child_list: bool,
+
+    /// If `true` it will observe all changes to the target's attributes.
+    pub attributes: bool,
+
+    /// If `true` it will observe all changes to the `CharacterData`'s data.
+    pub character_data: bool,
+
+    /// If `true` it will observe all changes to the target, the target's children, and the target's sub-children.
+    ///
+    /// This is recursive, so it causes **all** children and sub-children to be observed.
+    pub subtree: bool,
+
+    /// If `true` it will store the target's old attribute value in [`old_value`](enum.MutationRecord.html#variant.Attribute).
+    pub attribute_old_value: bool,
+
+    /// If `true` it will store the `CharacterData`'s old data in [`old_data`](enum.MutationRecord.html#variant.CharacterData).
+    pub character_data_old_value: bool,
+
+    /// If `Some` it will only observe the specified attributes. The attributes should be specified without a namespace.
+    ///
+    /// If `None` it will observe all attributes.
+    pub attribute_filter: Option< &'a [ &'a str ] >,
+}
+
+
+impl MutationObserver {
+    /// Returns a new `MutationObserver` with the given callback.
+    ///
+    /// The callback will be called with the following arguments when the observed DOM nodes change:
+    ///
+    /// 1. Vector of changes to the observed DOM nodes.
+    ///
+    /// 2. The `MutationObserver`.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver#Constructor)
+    pub fn new< F >( callback: F ) -> Self
+        where F: FnMut( Vec< MutationRecord >, Self ) + 'static {
+        js! (
+            return new MutationObserver( @{callback} );
```

Comment:
Rust's `Fn` and `FnMut` callbacks when passed to the `js!` macro must to be deallocated manually by calling `.drop()`. Right now any callback passed into `MutationObserver::new` will permanently leak memory, which is not good.

So, we need to figure out what to do with this. As far as I can see there are two cases we have to deal with:

1) We receive a `MutationObserver` created on the JavaScript side with a JavaScript callback. We treat this like any other `Reference` value.
2) We create our own `MutationObserver` with `new()`. We then know that we have a closure on Rust's side that needs to be deallocated.

One random idea that I have to deal with (2) would be modifying the `new` to return something like `OwnedMutationObserver` instead of `MutationObserver`. The `OwnedMutationObserver` would implement `Drop` where it would disconnect the observer and deallocate the callback. That way when the `OwnedMutationObserver` goes out of scope our callback would be automatically deallocated.

Reply:
If we have `new` return an `OwnedMutationObserver` which cleans things up when it goes out of scope, does the `FnMut` still need a `'static` lifetime?

Reply:
Yes, it still needs the `'static` lifetime; even if we'll make it automatically drop the callback when the `OwnedMutationObserver` goes out of scope the actual `OwnedMutationObserver` still could be, for example, assigned to some global variable and kept alive forever.

Reply:
But memory leaks are always possible in Rust, and that doesn't normally warrant `'static` lifetimes, so what's special about this situation? Since the reference will be held indefinitely in both JS and Rust, the memory should always be valid, right?

Reply:
As far as the `'static` lifetime is concerned it's not about memory leaks, it's about safety, and it's not really about the closure *itself*, but about what it's allowed to close over. In a nutshell the `'static` lifetime tells the compiler that any references passed into the closure by whoever creates it can only point to stuff with a lifetime of the whole program, so in particular they can *not* reference anything on the stack for example.

Technically we *could* make the closure have an arbitrary lifetime `'a` and then make the return type `OwnedMutationObserver< 'a >`, but it's probably not worth it.

Reply:
Ahh, right, the closure could hold references to a variable on the stack, which then won't exist anymore when the callback is called. I'll experiment with using `'a`, but you may be right that `'static` is necessary.